### ISSUE TITLE
Incomplete ShadowAddress (shadow for android.location.Address)

### ISSUE THREAD

**Issue by winniehell** [robolectric/robolectric#2182]
Right now, the following tests fail using Robolectric but do not fail on a device:

```java
    @Test
    public void testHasLatitude() {
        Address address = new Address(Locale.getDefault());
        address.setLatitude(42.0);
        Assert.assertEquals(true, address.hasLatitude());
    }

    @Test
    public void testHasLongitude() {
        Address address = new Address(Locale.getDefault());
        address.setLongitude(42.0);
        Assert.assertEquals(true, address.hasLongitude());
    }

    @Test
    public void testGetMaxAddressLineIndex() {
        Address address = new Address(Locale.getDefault());
        address.setAddressLine(0, "somewhere");
        Assert.assertEquals(0, address.getMaxAddressLineIndex());
    }

    @Test
    public void testGetAddressLine() {
        Address address = new Address(Locale.getDefault());
        address.setAddressLine(0, "somewhere");
        address.setAddressLine(1, "nowhere");
        Assert.assertEquals("somewhere", address.getAddressLine(0));
    }
```

Is this behavior intended?

**Comment by winniehell** [comment]
Would a pull request be welcome? :wink:

**Comment by SuperAmin** [comment]
I'm wondering why ShadowAddress exists at all since all of the getters and setters for android.location.Address are public and there are no Android specific dependencies in the class: http://grepcode.com/file/repository.grepcode.com/java/ext/com.google.android/android/5.1.1_r1/android/location/Address.java

Try deleting ShadowAddress and see if all of the tests pass (including yours).

**Comment by erd** [comment]
I'd be more than happy to look at a PR. Although, looking at the source for `Address`, it might be possible to just delete the whole shadow as @SuperAmin suggests.

**Comment by winniehell** [comment]
@SuperAmin: I don't understand how removing `ShadowAddress` could possibly work. The reason for its existence (and that of robolectric in general) is that the `android.jar` used for unit testing [contains only stubs][not-mocked]. As soon as you remove shadow classes, you'll get stuff like
```
java.lang.RuntimeException: Method setAddressLine in android.location.Address not mocked.
```

@erd: Do I have a large misunderstanding here?

[not-mocked]: http://g.co/androidstudio/not-mocked

**Comment by SuperAmin** [comment]
From the Robolectric website <http://robolectric.org/extending/>:
From Robolectric 2.0 on, the number of shadow classes needed is greatly
reduced, because real Android OS code is present. Methods on your shadow
class are able to call through to the Android OS code if you like, using
Robolectric.directlyOn().

**Comment by erd** [comment]
@winniehell Even though your code is compiled against the stubs in the SDK, Robolectric runs your tests against a version of Android built from AOSP. Most of Robolectric is code to enable the AOSP code to run on a desktop JVM. You really only need shadows for methods that are dangerous (i.e. Java code that calls into native code).

If the `Address` class in AOSP only contains getters and setters, it's probably safe to remove.

**Comment by winniehell** [comment]
@SuperAmin: Thanks, I did not know about that method. Looks like it is now `Shadow.directlyOn()`. It is located in `org.robolectric.internal.Shadow` and not documented&mdash;so it does not look like it is meant to be used to me. Additionally, it seems like the method [requires a shadow][Shadow-directlyOn].

@erd: That's the large misunderstanding I meant. I did not know that Robolectric is able to run Android code. Can you point me to example where this is done?

[Shadow-directlyOn]: https://github.com/robolectric/robolectric/blob/1021315c11e619e0735bea1349e04023178c4067/robolectric-utils/src/main/java/org/robolectric/internal/Shadow.java#L35

**Comment by erd** [comment]
`SdkConfig` declares the versions of Android that Robolectric supports. `InstrumentingClassLoader` is where we wire in calls to shadow classes when classes in the Android SDK are loaded.

`Shadow.directlyOn` is fine to use from a shadow, fwiw.

**Comment by erd** [comment]
This should be fixed by #2226.
